Source/MetroXHunter/Reload/ReloadComponent.cpp

```diff
@@ -0,0 +1,359 @@
+	float DeltaTime, 
+	ELevelTick TickType, 
+	FActorComponentTickFunction* ThisTickFunction 
+)
+{
+	Super::TickComponent( DeltaTime, TickType, ThisTickFunction );
+
+	ReloadElapsedTime += DeltaTime;
+	OnUpdateCursorPosition.Broadcast();
+
+	// If the reload has reached its normal duration, trigger a normal reload
+	if ( ReloadElapsedTime >= ReloadDataAsset->NormalReloadDuration )
+	{
+		SetComponentTickEnabled( false );
+
+		TriggerReload( EReloadState::Normal,
+			ReloadDataAsset->NormalReloadAnimTime,
+			ReloadDataAsset->NormalReloadAnimTime
+		);
+	}
+}
+
+void UReloadComponent::SetupPlayerInputComponent()
+{
+	if ( !PlayerController )
+	{
+		return;
+	}
+
+	UInputComponent* PlayerInputComponent = PlayerController->InputComponent;
+
+	// Set up action bindings
+	if ( UEnhancedInputComponent* EnhancedInputComponent = Cast<UEnhancedInputComponent>( PlayerInputComponent ) )
+	{
+		// Interaction
+		EnhancedInputComponent->BindAction( 
+			ReloadAction, ETriggerEvent::Started, 
+			this, &UReloadComponent::OnReloadInput 
+		);
+	}
+}
+
+void UReloadComponent::StartReloadSequence()
+{
+	if ( !bIsReloadActive && !IsGunFireLocked() )
+	{
+		bIsReloadActive = true;
+		CurrentGunState = EGunState::Reloading;
+		ReloadElapsedTime = 0.0f;
+
+		SetComponentTickEnabled( true );
+
+		// Notify that the reload has started
+		OnReloadStateChanged.Broadcast(EReloadState::Start);
+	}
+}
+
+void UReloadComponent::TriggerReload( EReloadState ReloadState, float ReloadAnimTime, float FinalWaitingTime )
+{
+	if ( bUseInfiniteAmmo )
+	{
+		FinalizeReload( MaxAmmoInMagazine, ReloadAnimTime, 0 );
+	}
+	else
+	{
+		int NewAmmoInMagazine;
+		int InventoryAmmoConsumed;
+
+		// Calculate the quantity of ammo to reload
+		ComputeReloadAmmoCount(NewAmmoInMagazine, InventoryAmmoConsumed);
+
+		FinalizeReload( NewAmmoInMagazine, ReloadAnimTime, InventoryAmmoConsumed );
+
+		OnReloadStateChanged.Broadcast(ReloadState);
+	}
+
+	GetWorld()->GetTimerManager().SetTimer(
+		TimerHandleReloadPlayback, [this]()
+		{ bIsReloadActive = false; },
+		FinalWaitingTime,
+		false
+	);
+}
+
+void UReloadComponent::UpdateCurrentGunState( EGunState NewState )
+{
+	CurrentGunState = NewState;
+}
+
+void UReloadComponent::RetrievePlayerInventory()
+{
+	AActor* Owner = GetOwner();
+	if ( Owner )
+	{
+		PlayerInventory = Cast<UInventoryComponent>( Owner->GetComponentByClass( UInventoryComponent::StaticClass() ) );
+	}
+}
+
+
+void UReloadComponent::RetrieveHUD()
+{
+	PlayerController = UGameplayStatics::GetPlayerController( this, 0 );
+	if ( PlayerController )
+	{
+		HUD = PlayerController->GetHUD();
+	}
+}
+
+void UReloadComponent::InitializeReloadData( UReloadData* NewReloadData )
+{
+	if ( NewReloadData )
+	{
+		ReloadDataAsset = NewReloadData;
+	}
+}
+
+void UReloadComponent::GetNormalizedReloadTimings( 
+	float& OutPerfectReloadStartTime, 
+	float& OutActiveReloadStartTime, 
+	float& OutActiveReloadEndTime 
+) const
+{
+	if ( ReloadDataAsset && ReloadDataAsset->NormalReloadDuration > 0.0f )
```

Comment:
I think it would fit more our code style to have this logic of "functions should fail as early as possible", in the sense where:
```cpp
if ( ReloadDataAsset == nullptr || ReloadDataAsset->NormalReloadDuration <= 0.0f )
{
    // set everything to 0.0f
    return;
}

// set everything to the correct values
```

It also avoid nesting